From ea6039a30929ff845859ed601594546d71894d84 Mon Sep 17 00:00:00 2001
From: Kay Sievers <kay@vrfy.org>
Date: Sun, 07 Oct 2012 16:21:38 +0000
Subject: udev: allow firmware requests to bypass the dependency tracking

The removal of the TIMEOUT= handling in udevd put firmware requests into the
devpath parent/child dependency tracking. Drivers which block in module_init()
asking userspace for firmware ran into a 30 sec device timeout.

The whole firmware loading willl hopefully move into the kernel and
the fragile-since-day-one fake async driver-core device dance involving
udev can be retired:
  http://git.kernel.org/?p=linux/kernel/git/torvalds/linux.git;a=commit;h=abb139e75c2cdbb955e840d6331cb5863e409d0e
--
--- udev-182/src/udevd.c.orig	2013-07-05 11:53:11.273415774 +0200
+++ udev-182/src/udevd.c	2013-07-05 11:53:26.517415954 +0200
@@ -103,8 +103,9 @@
         size_t devpath_len;
         const char *devpath_old;
         dev_t devnum;
-        bool is_block;
         int ifindex;
+        bool is_block;
+        bool nodelay;
 };
 
 static struct event *node_to_event(struct udev_list_node *node)
@@ -455,6 +456,9 @@
         event->is_block = (strcmp("block", udev_device_get_subsystem(dev)) == 0);
         event->ifindex = udev_device_get_ifindex(dev);
 
+        if (strcmp(udev_device_get_subsystem(dev), "firmware") == 0)
+                event->nodelay = true;
+
         udev_queue_export_device_queued(udev_queue_export, dev);
         info(event->udev, "seq %llu queued, '%s' '%s'\n", udev_device_get_seqnum(dev),
              udev_device_get_action(dev), udev_device_get_subsystem(dev));
@@ -542,6 +546,10 @@
                         return true;
                 }
 
+                /* allow to bypass the dependency tracking */
+                if (event->nodelay)
+                        continue;
+
                 /* parent device event found */
                 if (event->devpath[common] == '/') {
                         event->delaying_seqnum = loop_event->seqnum;
